### PR TITLE
fix(bootstrap): Add one time reset of user bootstrap node list

### PR DESCRIPTION
**Pull request by anthonybilinski**
- [x] Commits follow our [git commit guidelines](https://github.com/qTox/qTox/blob/master/CONTRIBUTING.md#git-commit-guidelines)

---
This change is [Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)

### CODE REVIEW

**Comment by sudden6** [comment]
do you think it makes sense to version the settings too? I have not seen this elsewhere, but I can't come up with a good reason not to do it.

**Review by anthonybilinski — COMMENTED** [review]
On reconsideration, yes I think that's best and I've changed it to work similar to `History`.

Upgrading is handled by Settings now too, rather than exposing the bug mitigation state to different classes.

> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465#-:-MxGKLiaEwQTPJZs27C_:b-g37uar)* status: 0 of 1 LGTMs obtained

File: src/persistence/settings.cpp [hunk 1]
```diff
@@ -112,16 +117,34 @@ void Settings::loadGlobal()
...
         filePath = ":/conf/" + globalSettingsFile;
+        defaultSettings = true;
     }
 
     qDebug() << "Loading settings from " + filePath;
 
     QSettings s(filePath, QSettings::IniFormat);
     s.setIniCodec("UTF-8");
 
+    s.beginGroup("Version");
```
Comment (sphaerophoria) [inline src/persistence/settings.cpp:132]: Why do you have a group with a single value? Is this just how ini's have to be formatted?

File: src/persistence/paths.cpp [hunk 2]
```diff
@@ -365,11 +365,25 @@ QString Paths::getAppCacheDirPath() const
 #endif
 }
 
+QString Paths::getExampleNodesFilePath() const
+{
+    QDir dir(getSettingsDirPath());
+    constexpr static char nodesFileName[] = "bootstrapNodes.example.json";
```
Comment (sphaerophoria) [inline src/persistence/paths.cpp:371]: This is a nice solution. Gives someone a reference without preventing us from updating it later.

File: src/persistence/globalsettingsupgrader.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+#include <QDebug>
+#include <QFile>
+
+namespace {
+    bool version0to1(Settings& settings)
+    {
+        const auto& paths = settings.getPaths();
+
+        QFile badFile{paths.getUserNodesFilePath()};
+        if (badFile.exists()) {
```
Comment (sphaerophoria) [inline src/persistence/globalsettingsupgrader.cpp:33]: Can we prompt the user for this? No one is going to see the qWarning

**Review by anthonybilinski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/settings.cpp, line 132 at r4](https://reviewable.io/reviews/qtox/qtox/6465#-MwP2HaK32e0anhnqnvN-r4-132:-MxHwF3A3xeenxck1AtO:b5xk9u4) ([raw file](https://github.com/qtox/qtox/blob/4febe25c27eeec2f9e5b49f0ca6bf70e97b77726/src/persistence/settings.cpp#L132)):*
<details><summary><i>Previously, sphaerophoria wrote…</i></summary><blockquote>

Why do you have a group with a single value? Is this just how ini's have to be formatted?
</blockquote></details>

It's semi-required by the file format, see bottom of this section: https://doc.qt.io/qt-5/qsettings.html#Format-enum

If it was not explicitly put in a group in code, it would be under "General" in the file, which doesn't seem much different.
___
*[src/persistence/globalsettingsupgrader.cpp, line 33 at r6](https://reviewable.io/reviews/qtox/qtox/6465#-MxGwIuv2LeMZ955T49Y-r6-33:-MxHwa-p9ciWcLznQZJH:b-sjmqje) ([raw file](https://github.com/qtox/qtox/blob/867c21b7fabdaa19dd0bc33e85d9fb4c177d9d12/src/persistence/globalsettingsupgrader.cpp#L33)):*
<details><summary><i>Previously, sphaerophoria wrote…</i></summary><blockquote>

Can we prompt the user for this? No one is going to see the qWarning
</blockquote></details>

We don't have a GUI at time of global settings load thanks to a mess of initialization order which wouldn't be easy to fix =/

I agree no one's going to see this initially, but they might look into logs or open an issue if qTox refuses to start, which this causes. Same issue in `Settings` where `GlobalSettingsUpgrader::doUpgrade` is called.

**Review by sudden6 — CHANGES_REQUESTED** [review]
> Reviewed 50 of 59 files at r4, 8 of 9 files at r7, 4 of 4 files at r8, 4 of 4 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/personalsettingsupgrader.cpp, line 26 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTuV0FBKdBaMY90J67:-MxTuV0FBKdBaMY90J68:btb4xej) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/personalsettingsupgrader.cpp#L26)):*
> ```cpp
> #include <cassert>
> 
> bool PersonalSettingsUpgrader::doUpgrade(SettingsSerializer& settingsSerializer, int fromVer, int toVer)
> ```

same, make these `unsigned`
___
*[src/persistence/settings.h, line 720 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTuZp34Ji8qdiwGvXt:-MxTuZp34Ji8qdiwGvXu:bb7od0r) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/settings.h#L720)):*
> ```objc
>     static QThread* settingsThread;
>     Paths paths;
>     int globalSettingsVersion = 0;
> ```

same
___
*[src/persistence/settings.cpp, line 493 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTupq7-uIiDhiDWpUh:-MxTupq7-uIiDhiDWpUi:b-r8h8u3) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/settings.cpp#L493)):*
> ```cpp
>     bool defaultSettings = false;
>     if (!QFile(filePath).exists()) {
>         defaultSettings = true;
> ```

move to the `else` path below
___
*[src/persistence/globalsettingsupgrader.cpp, line 26 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTuG8g4xn1wm9NeKy0:-MxTuG8g4xn1wm9NeKy1:b-3hznmd) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/globalsettingsupgrader.cpp#L26)):*
> ```cpp
> #include <cassert>
> 
> bool GlobalSettingsUpgrader::doUpgrade(Settings& settings, int fromVer, int toVer)
> ```

if you make these `unsigned int` you can probably remove the casts and I don't expect there will be negative versions.

**Review by sudden6 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/personalsettingsupgrader.cpp, line 26 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTuV0FBKdBaMY90J67:-MxUCciABGxTJHAVHc5x:b7caqk4) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/personalsettingsupgrader.cpp#L26)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

same, make these `unsigned`
</blockquote></details>

Retracted, based on out-of-band discussions with regards to Qt size types compatibility.

**Review by sudden6 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/settings.cpp, line 493 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTupq7-uIiDhiDWpUh:-MxUGzSX9pQJp_3Mh4U0:b-60qyt5) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/settings.cpp#L493)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

move to the `else` path below
</blockquote></details>

Sorry, I misread that and merging is not possible, but it can be converted to a oneliner: `const  bool defaultSettings = !QFile(filePath).exists()`

**Review by anthonybilinski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/settings.cpp, line 493 at r7](https://reviewable.io/reviews/qtox/qtox/6465#-MxTupq7-uIiDhiDWpUh:-MxUCeJJ128pCGa9Iw1J:b-1xgh2e) ([raw file](https://github.com/qtox/qtox/blob/77d7d4c96b2bf1011af6643565b2fce2f9ac76fd/src/persistence/settings.cpp#L493)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Sorry, I misread that and merging is not possible, but it can be converted to a oneliner: `const  bool defaultSettings = !QFile(filePath).exists()`
</blockquote></details>

arg I was thinking to avoid the case where they have a global settings file but no personal settings file, but this would miss every new profile for a single user. Running it only if it's a new profile settings file could potentially miss the upgrade if a user has a global settings file but no user settings file though.. which looks like a possible case based on this loading behaviour, and digging through history I found that that was in aea9eea8a4e2f95a7175954183959c1070bcc315. So it looks like we have no perfect way to tell if it's a new profile based on Settings files.

That should really be done through an upgrade itself at some point...

For now though, I'll just pass in if it's a new profile from Profile to Settings::loadPersonal, since Profile is aware and is always doing the loading.

**Review by sudden6 — APPROVED** [review]
:lgtm_strong: 

> Reviewed 8 of 8 files at r10, 4 of 4 files at r11, 4 of 4 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/6465#-:-MxUdsb92tIB_0zF47vf:b-uwkowd)* status: :shipit: complete! 1 of 1 LGTMs obtained

**Pull request merged**